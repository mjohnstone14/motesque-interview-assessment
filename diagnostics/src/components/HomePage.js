import React, { useContext } from 'react';
import '../styles/App.css';
import CategoryAccordion from '../base-components/CategoryAccordion';
import AllTests from '../base-components/AllTests';
import AppContext from '../context/AppContext';
import { Typography } from '@material-ui/core';
import { ToastContainer } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

/**
 * Component that lists the test buttons in accordions based on category.
 * Each button is constructed to have a fetch POST onClick to the API
 */
export default function HomePage() {
  const context = useContext(AppContext);

  /**
   * Appends category accordions that map
   * respective test button components to each
   */
  const renderCategoryAccordions = () => {
    let categories = context.getCategories();
    let testAccordions = [];
    categories.forEach(async (category) => {
      context.setCategoryTests(category);
      testAccordions.push(
        <CategoryAccordion
         key={category.category}
         title={category.category}
         detail={category.description}
        />
      );
    });
    return testAccordions;
  };

  return (
    <AppContext.Consumer>
      {() => (
        <div>
          <ToastContainer 
            position="top-right"
            autoClose={2500}
            hideProgressBar={false}
            newestOnTop={false}
            closeOnClick
            rtl={false}
            pauseOnFocusLoss
            draggable
            pauseOnHover
          />
          <Typography variant='h3' className="App-title-style">
            Motesque Test Executor
          </Typography>
          {(!context.allTestsRetrievedFlag()) ?
            <AllTests 
              title="Get All Tests" 
            /> : renderCategoryAccordions()
          }
        </div>
      )}
    </AppContext.Consumer>  
  )
};