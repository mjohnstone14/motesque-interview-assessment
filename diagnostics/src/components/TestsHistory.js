import React, { useContext } from 'react';
import AppContext from '../context/AppContext';
import { Typography } from '@material-ui/core';
import Accordion from '@material-ui/core/Accordion';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import AccordionSummary from '@material-ui/core/AccordionSummary';
import Paper from '@material-ui/core/Paper';
import '../styles/App.css';

export default function TestsHistory() {
  const context = useContext(AppContext);

  /**
   * Iterates over the map of results and pushes styled elements
   * to a list of elements to be displayed within an accordion.
   * The map contains a k,v pair of (category, -1) to help organize
   * the elements under their respective category title
   */
  const renderHistory = () => {
    const historyElements = [];
    let historyMap = context.getUpdatedHistory();
    // print history map
    console.log(historyMap);
    // iterate over key value pairs and determine category
    historyMap.forEach((value, key) => {
      if (key === 'System' || key === 'Sensor' || key === 'Specific') {
        historyElements.push(
          <Typography
            key={'title' + key}
            style={{display: 'flex', marginLeft: '2%', fontStyle:'italic'}} 
            variant='h6'>Number of {key} Test Runs
          </Typography>
        )
      } else {
        historyElements.push(
          <Paper key={key} className="PaperTestResultStyle">
            <Typography key={'listItem' + key} style={{
              display:'flex',
              marginRight: '1%', 
            }}>
                {key}:
            </Typography>
            <Typography key={'listItem' + value} style={{
              display: 'flex',
              textTransform: 'capitalize',
              color:'#ff6a00',
              fontWeight: 'bold'}}>
                {value}
            </Typography>
          </Paper>
        )
      }
    });
    return historyElements;
  }
  
  return (
    <AppContext.Consumer>
      {() => (
        <div>
          <Typography
            key='historyHeader'
            className="App-title-style"
            variant='h4' 
            style={{marginBottom: '0.8%'}}
          >
            Test History
          </Typography>
          <Accordion key='testHistoryAccordion' style={{borderRadius: '25px'}}>
            <AccordionSummary
              expandIcon={<ExpandMoreIcon />}
              aria-controls="panel1a-content"
              id="panel1a-header"
              style={{fontWeight: 'bold', textTransform: 'capitalize'}}
            >
              View all tests passed
            </AccordionSummary>
            {renderHistory()}
          </Accordion>
        </div>
      )}
    </AppContext.Consumer>  
  )
};