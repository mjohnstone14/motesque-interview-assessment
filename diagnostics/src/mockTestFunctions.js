
  export const allTestsRetrievedFlag = jest.fn(() => true);
  export const setAllTestsRetrievedFlag = jest.fn();
  export const setCategories = jest.fn();
  export const getCategories = jest.fn(() => {
    return {
        "tests": [
          {
            "category": "system",
            "description": "check system",
            "tests": [
              {
                "name": "Check system information",
                "route": "/system/check-info"
              },
              {
                "name": "Check system free storage",
                "route": "/system/check-free-storage"
              },
              {
                "name": "Check database status",
                "route": "/system/check-database-status"
              },
              {
                "name": "Check USB ports",
                "route": "/system/check-usb-ports"
              }
            ]
          },
          {
            "category": "sensor",
            "description": "check sensors",
            "tests": [
              {
                "name": "Check sensor firmware",
                "route": "/sensor/firmware-matches-base-station-version"
              },
              {
                "name": "Check sensor hardware",
                "route": "/sensor/check-hardware"
              },
              {
                "name": "Check sensor calibration",
                "route": "/sensor/check-calibration"
              },
              {
                "name": "Run test measurement",
                "route": "/sensor/measurement/test"
              },
              {
                "name": "Test sensor battery",
                "route": "/sensor/battery/test"
              }
            ]
          },
          {
            "category": "specific",
            "description": "check customer",
            "tests": [
              {
                "name": "Get customer",
                "route": "/specific/customer"
              },
              {
                "name": "Check running services",
                "route": "/specific/check-running-services"
              },
              {
                "name": "Test API",
                "route": "/specific/api/test"
              }
            ]
          }
        ]
      }
  });