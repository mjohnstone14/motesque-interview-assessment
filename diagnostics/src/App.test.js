import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'
import HomePage from './components/HomePage';
import TestsHistory from './components/TestsHistory';
import AppContext from './context/AppContext';
import * as mockFunctions from './mockTestFunctions';

test('HomePage matches snapshot', () => {
  const allTestsRetrievedFlag = mockFunctions.allTestsRetrievedFlag;
  const getCategories = mockFunctions.getCategories;
  const setCategories = mockFunctions.setCategories;
  const setAllTestsRetrievedFlag = mockFunctions.setAllTestsRetrievedFlag;
  const { container } = render(
    <AppContext.Provider
     value={{
      allTestsRetrievedFlag,
      getCategories,
      setCategories,
      setAllTestsRetrievedFlag
    }}>
      <HomePage />
    </AppContext.Provider>);
  expect(container.firstChild).toMatchSnapshot();
})

test('TestsHistory matches snapshot', () => {
  const getUpdatedHistory = jest.fn(() => {
    return new Map([
      ['System', -1],
      ['Check system information', 5],
      ['Check system free storage', 4],
      ['Check database status', 3],
      ['Check USB ports', 2],
      //sensors
      ['Sensor', -1],
      ['Check sensor firmware', 8],
      ['Check sensor hardware', 9],
      ['Check sensor calibration', 1],
      ['Run test measurement', 2],
      ['Test sensor battery', 0],
      //specific
      ['Specific', -1],
      ['Get customer', 4],
      ['Check running services', 2],
      ['Test API', 0]
    ])
  });
  const { container } = render(
    <AppContext.Provider value={{getUpdatedHistory}}>
      <TestsHistory />
    </AppContext.Provider>);
  expect(container.firstChild).toMatchSnapshot();
})

