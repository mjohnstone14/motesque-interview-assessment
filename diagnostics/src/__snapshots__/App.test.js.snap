// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches snapshot 1`] = `
<div>
  <div
    class="Toastify"
  />
  <h3
    class="MuiTypography-root App-title-style MuiTypography-h3"
  >
    Motesque Test Executor
  </h3>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    style="margin: 5%;"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Get All Tests
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`TestsHistory matches snapshot 1`] = `
<div>
  <h4
    class="MuiTypography-root App-title-style MuiTypography-h4"
    style="margin-bottom: 0.8%;"
  >
    Test History
  </h4>
  <div
    class="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
    style="border-radius: 25px;"
  >
    <div
      aria-controls="panel1a-content"
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root"
      id="panel1a-header"
      role="button"
      style="font-weight: bold; text-transform: capitalize;"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content"
      >
        View all tests passed
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel1a-header"
            id="panel1a-content"
            role="region"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6"
              style="display: flex; margin-left: 2%; font-style: italic;"
            >
              Number of 
              System
               Test Runs
            </h6>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check system information
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                5
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check system free storage
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                4
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check database status
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                3
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check USB ports
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                2
              </p>
            </div>
            <h6
              class="MuiTypography-root MuiTypography-h6"
              style="display: flex; margin-left: 2%; font-style: italic;"
            >
              Number of 
              Sensor
               Test Runs
            </h6>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check sensor firmware
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                8
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check sensor hardware
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                9
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check sensor calibration
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                1
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Run test measurement
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                2
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Test sensor battery
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                0
              </p>
            </div>
            <h6
              class="MuiTypography-root MuiTypography-h6"
              style="display: flex; margin-left: 2%; font-style: italic;"
            >
              Number of 
              Specific
               Test Runs
            </h6>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Get customer
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                4
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Check running services
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                2
              </p>
            </div>
            <div
              class="MuiPaper-root PaperTestResultStyle MuiPaper-elevation1 MuiPaper-rounded"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; margin-right: 1%;"
              >
                Test API
                :
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
                style="display: flex; text-transform: capitalize; color: rgb(255, 106, 0); font-weight: bold;"
              >
                0
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
